### PR TITLE
InstData: Generator commands to create entities and source files tsv

### PR DESCRIPTION
**Pull request by ebelter**
This PR includes 2 new commands added to 'genome instrument-data import'. Both start from a modified spreadsheet. The goal for these commands is to have analysts and data transfer be able to run imports  on their own. Confluence doc: https://confluence.gsc.wustl.edu/display/BIO/GMS+-+Importing+Instrument+Data
generate-entity-create-commands => This replaces the sample importers. It outputs commands to create all the entities needed to import instrument data. 
generate-source-files-tsv =>  This generates the source files tsv. This file is used in the import manager.

### CODE REVIEW

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+use Params::Validate qw( :types );
+use Text::CSV;
+
+class Genome::InstrumentData::Command::Import::GenerateBase {
+    is => 'Command::V2',
+    is_abstract => 1,
+    has_input => {
+        file => {
+            is => 'Text',
+            doc => 'Comma (.csv) or tab (.tsv) separated file of entity names, attributes and other meta data. Separartor is determined by file extension.',
```
Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: typo s/Separartor/Separator

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+ Required
+  sample.name_part => Name or id from external source. If name is given, the individual and library names will be resolved from it.
+  sample.extraction_type => 'genomic dna' or 'rna'
+
+ Optional, but recommended:
+  sample.common_name    => Usually normal or tumor to indicate disease state.
+
+Library\n
+ Optional
+  library.ext           => Extension to append to the sample name. Deault is 'extlibs'.
```
Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: Typo s/Deault/Default

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    return (qw/ individual sample library instdata/);
+}
+
+sub _open_file_parser {
+    my $self = shift;
+
+    my $file = $self->file;
+    my ($dir, $basename, $ext) = File::Basename::fileparse($file, 'csv', 'tsv');
+    die $self->error_message("Cannot determine type for file: %s. It needs to end with .csv or .tsv.", $file) if not $ext;
+    my $parser = Text::CSV->new({
```
Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: Is there a reason not to use Genome::Utility::IO::SeparatedValueReader?

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateEntityCreateCommands.pm [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+sub _add_library_create_command {
+    my ($self, $entity_params) = Params::Validate::validate_pos(@_, {type => HASHREF}, {type => HASHREF});
+    my $params = $entity_params->{library};
+    $params->{sample} = $entity_params->{sample}->{name}; # add the sample name to the params
+    return $self->_add_create_command_for_type('library', $params);
+}
+
+sub _add_create_command_for_type {
+    my ($self, $type, $params, $dependent_params) = Params::Validate::validate_pos(
+        @_, {type => HASHREF}, {type => SCALAR}, {type => HASHREF},# {type => HASHREF},
```
Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateEntityCreateCommands.pm]: Why is the last bit commented out?

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    my $nomenclature = delete $entity_params->{sample}->{nomenclature};
+    my $individual_name_part = delete $entity_params->{individual}->{name_part};
+    my $sample_name_part = delete $entity_params->{sample}->{name_part};
+
+    # sample
+    my $sample_name = $entity_params->{sample}->{name};
+    if ( $sample_name ) {
+        my @tokens = split(/\-/, $sample_name);
+        if ( @tokens <  3 ) {
+            die $self->error_message('Invalid sample name: %s. It must have at least 3 parts separated by dashes.', $sample_name);
```
Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: If this is an existing sample in the db, is it possible that the sample name would not conform to the spec?

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        _entity_attributes => { is => 'ARRAY', },
+        _nomenclature => { is => 'Text', },
+        _names_seen => { is => 'HASH', default_value => {}, },
+    }
+};
+
+sub help_detail {
+    return <<HELP;
+The file should be a comma or tab separated values and indicated with the appropriate extension (csv and tsv). Column headers to use to generate the create commands should start with the entity (individual, sample, library, instdata) name then a period (.) and then then attribute name (Ex: sample.name_part). Here are some required and optional columns. For more, see each entity's create command (Ex: genome sample create --h). Please see Confluence documentation for more information and a full example.
+
```
Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: The description of the column headers doesn't match what is in the wiki page

Comment (apregier) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: Maybe I am confused by the wiki page - why does the example spreadsheet have BioSample_s ?

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm (hunk 5 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: Sure. How but would we know that?

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm (hunk 6 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: The wiki has a real life example from an SRA metadata file. I guess the _s means string? I need to add the finished spreadsheet to the wiki. Did I answer your Q?

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm (hunk 3 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateBase.pm]: I might have hasd a reason, but forgot. I guess I didn't see the SVR adding any value here. It should use Text::CSV under the hood. I can replace it if necessary.

File: lib/perl/Genome/InstrumentData/Command/Import/GenerateEntityCreateCommands.pm (hunk 4 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/GenerateEntityCreateCommands.pm]: yep

**Comment by ebelter** [comment]
Basic fixes ready, but waiting until further comment from @apregier.

**Comment by apregier** [comment]
+1

**Pull request merged**